skbio/tree/_upgma.py

```diff
@@ -0,0 +1,90 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2013--, scikit-bio development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file LICENSE.txt, distributed with this software.
+# ----------------------------------------------------------------------------
+
+import numpy as np
+
+from scipy.cluster.hierarchy import average, weighted
+from skbio import TreeNode
+from skbio.stats.distance import DistanceMatrix
+
+
+def upgma(dm: DistanceMatrix, weighted=False) -> TreeNode:
```

Comment:
Here, you are performing a very good practice: [typing](https://docs.python.org/3/library/typing.html). However, scikit-bio is backward compatible with as early as Python 3.8, and typing wasn't widely adopted by the time. Therefore, scikit-bio so far doesn't universally have typing in its codebase. Therefore we may skip typing until further discussions.